cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -534,47 +546,126 @@ public Void call() throws Exception {
+          String.format("Type mismatch in output type of mappers: %s and %s. " +
+                          "Map output key types: %s and %s. " +
+                          "Map output value types: %s and %s.",
+                        firstMapperClass, mapperClass,
+                        firstMapOutputKey, mapOutputKey,
+                        firstMapOutputValue, mapOutputValue));
+      }
+    }
+    return firstType;
+  }
+
+  /**
+   * Sets the configurations used for inputs.
+   */
+  private void setInputsIfNeeded(Job job) throws IOException {
+    Map<String, MapperInput> mapperInputs = context.getMapperInputs();
+
+    if (mapperInputs.size() == 0) {
+      // user is not going through our APIs to add input; leave the job's input format to user
+      return;
+    }
+
+    // keep track of which mapper was defined directly on the job, to use as the default
+    MultipleInputs.setRootMapper(job);
+    // TypeToken of the job set on the job:
+    TypeToken<Mapper> rootMapperTypeToken =
+      resolveClass(job.getConfiguration(), MRJobConfig.MAP_CLASS_ATTR, Mapper.class);
+
+    for (Map.Entry<String, MapperInput> mapperInputEntry : mapperInputs.entrySet()) {
+      MapperInput mapperInput = mapperInputEntry.getValue();
+      InputFormatProvider provider = mapperInput.getInputFormatProvider();
+      Map<String, String> inputFormatConfiguration = new HashMap<>(provider.getInputFormatConfiguration());
+
+      String inputName = mapperInputEntry.getKey();
+      Class<? extends Mapper> mapperClass = mapperInput.getMapper();
 
       // A bit hacky for stream.
-      // For stream, we need to do two extra steps.
-      // 1. stream usage registration since it only happens on client side.
-      // 2. Infer the stream event decoder from Mapper/Reducer
       if (provider instanceof StreamInputFormatProvider) {
-        StreamInputFormatProvider streamProvider = (StreamInputFormatProvider) provider;
-        Type inputValueType = getInputValueType(jobConf, StreamEvent.class);
-        ConfigurationUtil.setAll(streamProvider.setDecoderType(new HashMap<String, String>(), inputValueType), jobConf);
-
-        Id.Stream streamId = streamProvider.getStreamId();
-        try {
-          usageRegistry.register(context.getProgram().getId(), streamId);
-          streamAdmin.addAccess(new Id.Run(context.getProgram().getId(), context.getRunId().getId()),
-                                streamId, AccessType.READ);
-        } catch (Exception e) {
-          LOG.warn("Failed to register usage {} -> {}", context.getProgram().getId(), streamId, e);
-        }
+        setDecoderForStream((StreamInputFormatProvider) provider, job, inputFormatConfiguration,
+                            mapperClass == null ? rootMapperTypeToken : resolveClass(mapperClass, Mapper.class));
       }
+
+      MultipleInputs.addInput(job, inputName,
+                              provider.getInputFormatClassName(), inputFormatConfiguration, mapperClass);
+    }
```

Comment:
yea, both methods iterate over the inputs, but I think keeping the implementation in separate methods (because they're doing separate things) is cleaner.
One iteration does the type checking, and other iteration does setting of input configuration.
I think its cleaner like it is.